Timeout

paul-ollis opened the pull request:
FYI: Use of OS timers to timeout syntax highlighting and searching.

Removes the need to regularly read any system timers while executing the RE engine.

paul-ollis committed:
Use timeout functions for slow REs.

paul-ollis committed:
Clean up more vim_regexec_multi calls.

paul-ollis committed:
Fix for non-FEAT_RELTIME builds.

Dropped attempt to use const timed_out flag. Rmoves some more
conditional code, so I am quite happy.

paul-ollis committed:
Squash unused variable error.

puremourning commented on src/os_unix.c:
```diff
@@ -8256,3 +8262,129 @@ xsmp_close(void)
+#ifdef FEAT_RELTIME
+static struct itimerval prev_interval;
+static struct sigaction prev_sigaction;
+static int timeout_flag         = FALSE;
```
is it risky that this is global? i.e. if start_timeout is called-reentrantly? 

is it possible for the regex code to be re-entrant? A lot of vim code has to handle this sort of thing.

paul-ollis replied:
A valid question.

I did review the code to check for such issues and found none. Not really surprising since this is replacing one, single global timeout mechanism for another.

paul-ollis committed:
Fixes for Windows test failures.

paul-ollis committed:
Try to make search timeout test robust.

Test was failing due, I think, to insufficently large time bands within
the assertions.

paul-ollis committed:
Make tests work around Windows 7 timer issue.

Windows 7 can expire timers early! So this adjusts the values used in
some assertions.

paul-ollis committed:
Allow even more leeway for search timeout test.

paul-ollis committed:
Avoid setitimer for Unix and MacOS.

Timeouts using setitimer are still supported as a fallback, since that
may work for other Posix platforms.

paul-ollis committed:
Squash unused argument errors.
